packages/flutter_tools/test/general.shard/fake_process_manager_test.dart

```diff
@@ -0,0 +1,175 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'package:fake_async/fake_async.dart';
+
+import '../src/common.dart';
+import '../src/fake_process_manager.dart';
+
+void main() {
+  group(FakeProcess, () {
+    testWithoutContext('exits with specified exit code', () async {
+      final FakeProcess process = FakeProcess(exitCode: 42);
+      expect(await process.exitCode, 42);
+    });
+
+    testWithoutContext('exits with specified stderr, stdout', () async {
+      final FakeProcess process = FakeProcess(
+        stderr: 'stderr\u{FFFD}'.codeUnits,
+        stdout: 'stdout\u{FFFD}'.codeUnits,
+      );
+      await process.exitCode;
+
+      // Verify that no encoding changes have been applied to output.
+      //
+      // In the past, we had hardcoded UTF-8 encoding for these streams in
+      // FakeProcess. When a specific encoding is desired, it can be specified
+      // on FakeCommand or in the encoding parameter of FakeProcessManager.run
+      // or FakeProcessManager.runAsync.
+      expect((await process.stderr.toList()).expand((List<int> x) => x), 'stderr\u{FFFD}'.codeUnits);
+      expect((await process.stdout.toList()).expand((List<int> x) => x), 'stdout\u{FFFD}'.codeUnits);
+    });
+
+    testWithoutContext('exits after specified delay (if no completer specified)', () {
+      final bool done = FakeAsync().run<bool>((FakeAsync time) {
+        final FakeProcess process = FakeProcess(
+          duration: const Duration(seconds: 30),
+        );
+
+        bool hasExited = false;
+        unawaited(process.exitCode.then((int _) { hasExited = true; }));
+
+        // Verify process hasn't exited before specified delay.
+        time.elapse(const Duration(seconds: 15));
+        expect(hasExited, isFalse);
+
+        // Verify process has exited after specified delay.
+        time.elapse(const Duration(seconds: 20));
+        expect(hasExited, isTrue);
+
+        return true;
+      });
+      expect(done, isTrue);
+    });
+
+    testWithoutContext('exits when completer completes (if no duration specified)', () {
+      final bool done = FakeAsync().run<bool>((FakeAsync time) {
+        final Completer<void> completer = Completer<void>();
+        final FakeProcess process = FakeProcess(
+          completer: completer,
+        );
+
+        bool hasExited = false;
+        unawaited(process.exitCode.then((int _) { hasExited = true; }));
+
+        // Verify process hasn't exited when all async tasks flushed.
+        time.elapse(Duration.zero);
```

Comment:
Could/should this also be `time.flushMicrotasks();`? It'd be nice if async task flushing was consistent for the before and after cases. I don't feel strongly about this, feel free to skip

Reply:
Oh nice! This is a great question with a fun answer!

In Dart, there are two event queues:
* The microtask queue, which is flushed at the beginning of each turn through the main event loop. Microtasks are scheduled with [`scheduleMicrotask`](https://api.flutter.dev/flutter/dart-async/scheduleMicrotask.html).
* The main event queue, from which we pull regular async tasks. Regular tasks are scheduled via Timers, Futures, and async blocks which are just syntactic sugar for Futures.

Unfortunately, I'm not aware of a great modern reference to this on the Flutter or Dart sites, but this [archived article](https://web.archive.org/web/20170704074724/https://webdev.dartlang.org/articles/performance/event-loop) off the old dartlang site covers in a slightly more web-oriented context.

`FakeAsync.flushMicrotasks` will flush the microtask queue but doesn't guarantee a full spin through the main event queue. In this case, we're waiting for a task queued up with a delay on the main event queue, so we need to be sure that all tasks, not just microtasks, are flushed. I don't actually see it documented in the Completer.complete documentation, but my recollection is that Completers complete to the microtask queue, which bumps them to the head of the list of tasks to process, which is why we can get away with a flushMicrotasks call below. I should probably change the flushMicrotasks to just await, to force a turn through the event loop.